gluten-core/src/main/scala/io/glutenproject/utils/PullOutProjectHelper.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.utils
+
+import io.glutenproject.extension.columnar.TransformHints
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.execution.SparkPlan
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.collection.mutable
+
+trait PullOutProjectHelper {
+
+  val generatedNameIndex = new AtomicInteger(0)
+
+  /** The majority of Expressions only support Attribute when converting them into native plans. */
+  protected def isNotAttribute(expression: Expression): Boolean = expression match {
+    case _: Attribute => false
+    case _ => true
+  }
+
+  protected def getProjectExpressionMap = new mutable.HashMap[ExpressionEquals, NamedExpression]()
+
+  protected def getAndReplaceProjectAttribute(
+      expr: Expression,
+      projectExprsMap: mutable.HashMap[ExpressionEquals, NamedExpression]): Expression =
+    expr match {
+      case alias: Alias =>
+        projectExprsMap.getOrElseUpdate(ExpressionEquals(alias.child), alias).toAttribute
```

Comment:
do we need introduce `ExpressionEquals`? it is only be used in vanilla spark common subexpression elimination. I think `expr.canonicalized` is enough.